### PR TITLE
Adding TransformersNlpEngine

### PR DESCRIPTION
**Pull request by erann1987**
## Change Description

Added the ability to support HaggingFace Transformers, for Name Entity Recognition task, integrated in spacy pipeline.

## Issue reference

This PR fixes issue #887 

## Checklist

- [X] I have reviewed the [contribution guidelines](https://github.com/microsoft/presidio/blob/main/CONTRIBUTING.md)
- [X] I have signed the CLA
- [X] My code includes unit tests
- [X] All unit tests and lint checks pass locally
- [] My PR contains documentation updates / additions if required

### CODE REVIEW

**Review by omri374 — COMMENTED** [review]
Great work! added some comments but overall looks great

**Comment by omri374** [comment]
@erann1987 there's a test failing due to a missing import: https://dev.azure.com/csedevil/Presidio/_build/results?buildId=14352&view=logs&j=c159508c-3a29-50e1-f978-bea95db34528&t=b8c9e062-1bed-527e-b318-fc1a13efd07a&l=35

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/transformers_recognizer.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
+from typing import Tuple
+from collections import Set
```
Comment (omri374) [inline presidio-analyzer/presidio_analyzer/predefined_recognizers/transformers_recognizer.py:2]: should be `from typing import Set` (could be merged with the first line)

File: presidio-analyzer/presidio_analyzer/nlp_engine/transformers_nlp_engine.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+        ):
+            logger.error(
+                "'models.model_name' must contains 'spacy' and 'transformers' keys"
+            )
+            raise KeyError(
+                "Expected keys ('spacy' and 'transformers') was not found in "
+                "models.model_name dict"
+            )
+
+        # TODO: add defaults and input validation
```
Comment (omri374) [inline presidio-analyzer/presidio_analyzer/nlp_engine/transformers_nlp_engine.py:135]: Please remove TODO, already done

File: presidio-analyzer/presidio_analyzer/nlp_engine/transformers_nlp_engine.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        logger.debug(f"Loading SpaCy models: {models.values()}")
```
Comment (omri374) [inline presidio-analyzer/presidio_analyzer/nlp_engine/transformers_nlp_engine.py:136]:
```suggestion
        logger.debug(f"Loading SpaCy and transformers models: {models.values()}")
```

File: presidio-analyzer/presidio_analyzer/predefined_recognizers/transformers_recognizer.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+
+class TransformersRecognizer(SpacyRecognizer):
+    """
+    Recognize entities using the transformers package.
+
+    See https://huggingface.co/docs/transformers/main/en/index
+    Uses the transformers package
+    (https://huggingface.co/docs/transformers/main/en/installation) to align
+    transformers interface with spaCy
```
Comment (omri374) [inline presidio-analyzer/presidio_analyzer/predefined_recognizers/transformers_recognizer.py:16]: Worth mentioning that this recognizer doesn't run transformers models, but loads the output from the NlpArtifacts

**Review by omri374 — APPROVED** [review]

File: presidio-analyzer/conf/transformers.yaml [hunk 5]
```diff
@@ -0,0 +1,7 @@
+nlp_engine_name: transformers
+models:
+  -
```
Comment (shiranr) [inline presidio-analyzer/conf/transformers.yaml:3]: Need to fix this yaml.

File: presidio-analyzer/tests/test_nlp_engine_provider.py [hunk 6]
```diff
@@ -8,6 +8,7 @@
     StanzaNlpEngine,
     NlpEngineProvider,
 )
+from presidio_analyzer.nlp_engine.transformers_nlp_engine import TransformersNlpEngine
```
Comment (shiranr) [inline presidio-analyzer/tests/test_nlp_engine_provider.py:13]: What about E2E tests for the new engine?

**Review by shiranr — COMMENTED** [review]
Left 2 comments

**Review by shiranr — CHANGES_REQUESTED** [review]
Left 2 comments

File: presidio-analyzer/tests/test_nlp_engine_provider.py (hunk 6 above)
Comment (omri374) [reply presidio-analyzer/tests/test_nlp_engine_provider.py:13]: Totally agree. Current tests assume the package (transformers) is not installed, as this is a different flavor (pip install presidio-analyzer[transformers]. 
We should think about e2e in addition to unit tests that do install the package, potentially on a parallel pipeline. WDYT @shiranr @erann1987?

File: presidio-analyzer/conf/transformers.yaml (hunk 5 above)
Comment (omri374) [reply presidio-analyzer/conf/transformers.yaml:3]: What's wrong with it? Checked with YAML validator and it's valid

**Comment by omri374** [comment]
Points for a next PR to complete this:
1. end to end tests (current unit tests focus on input, but not on actually running the models)
2. Injection of the transformer models into the docker container (update `install_nlp_models.py` to include downloading a transformer model)

File: presidio-analyzer/conf/transformers.yaml (hunk 5 above)
Comment (shiranr) [reply presidio-analyzer/conf/transformers.yaml:3]: might be wrong but there is a line and the content one line after...

Comment (shiranr) [reply presidio-analyzer/conf/transformers.yaml:3]: bur if it's valid I'll resolve :)

File: presidio-analyzer/tests/test_nlp_engine_provider.py (hunk 6 above)
Comment (shiranr) [reply presidio-analyzer/tests/test_nlp_engine_provider.py:13]: sounds good. do you want to do it in this PR or another one?

Comment (omri374) [reply presidio-analyzer/tests/test_nlp_engine_provider.py:13]: As this isn't part of the critical path and only adds new capability, I would suggest to finish the PR and do another solely for e2e tests

Comment (omri374) [reply presidio-analyzer/tests/test_nlp_engine_provider.py:13]: There is also the docker integration to add before doing e2e tests

**Review by shiranr — APPROVED** [review]

**Comment by tradicio** [comment]
Hi to everyone! I tried to use TransformersNlpEngine for a multilingual version of presidio-analyzer but I realized that the last update of the pypi repository does not include this engine provider. Do you know when the next pypi release is planned? TransformersNlpEngine will be included in it? Thanks a lot for your hard work!

**Comment by omri374** [comment]
Hi @tradicio, you are correct, it is still not released. We have several small bugs related to it. Hope to have those fixed and a release a new version in the next month or so. 
In the meantime, you could consider cloning the repo.

**Pull request merged**